File: src/CmdPanel.cpp
// CmdPanel.cpp - This file is part of QPlot

/* QPlot - Publication quality 2D graphs with dual coordinate systems
   Copyright (C) 2014  Daniel Wagenaar
  
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// CmdPanel.C

#include "CmdPanel.h"


static CBuilder<CmdPanel> cbPanel("panel");

bool CmdPanel::usage() {
  return error("Usage: panel ID [x y w h] | -");
}

bool CmdPanel::parse(Statement const &s) {
  if (s.length()==2 && s[1].typ==Token::DASH)
    return true;
  else if (s.length()==2 &&
	   (s[1].typ==Token::CAPITAL || s[1].typ==Token::BAREWORD))
    return true;
  else if (s.length()==2+4 &&
	   (s[1].typ==Token::CAPITAL || s[1].typ==Token::BAREWORD) &&
	   s[2].typ==Token::NUMBER &&
	   s[3].typ==Token::NUMBER &&
	   s[4].typ==Token::NUMBER &&
	   s[5].typ==Token::NUMBER)
    return true;
  else
    return usage();
}


void CmdPanel::render(Statement const &s, Figure &f, bool dryrun) {
  f.leavePanel();
  if (s[1].typ==Token::DASH)
    return;
  
  QRectF area;
  if (s.length()==6) {
    area = QRectF(pt2iu(s[2].num), pt2iu(s[3].num),
		  pt2iu(s[4].num), pt2iu(s[5].num));
    if (!dryrun)
      f.painter().drawRect(area);
  }
  f.choosePanel(s[1].str);
  if (s.length()==6)
    f.setExtent(area);
}

File: backend/python/cmdpanel.py
# cmdpanel.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems
# Copyright (C) 2014  Daniel Wagenaar
# (same GPL licence as original)

from __future__ import annotations

from typing import TYPE_CHECKING

from PyQt6.QtCore import QRectF

from .command import Commandfrom .token_ import Tokenfrom .factor import pt2iu
if TYPE_CHECKING:
    from .statement import Statement    from .figure import Figure

@Command.register("panel")
class CmdPanel(Command):
    """Select or create a panel.

    Syntax:
        panel -               (leave current panel, select none)
        panel ID              (select panel by ID)
        panel ID x y w h      (select panel and set its extent in pt)
    """

    def _usage(self) -> bool:
        return self.error("Usage: panel ID [x y w h] | -")

    def parse(self, s: Statement) -> bool:
        if len(s) == 2 and s[1].typ == Token.DASH:
            return True
        if (len(s) == 2
                and s[1].typ in (Token.CAPITAL, Token.BAREWORD)):
            return True
        if (len(s) == 6
                and s[1].typ in (Token.CAPITAL, Token.BAREWORD)
                and all(s[k].typ == Token.NUMBER for k in range(2, 6))):
            return True
        return self._usage()

    def render(self, s: Statement, f: Figure, dryrun: bool) -> None:
        f.leave_panel()
        if s[1].typ == Token.DASH:
            return
        f.choose_panel(s[1].str)
        if len(s) < 6:
            return
        area = QRectF(
            pt2iu(s[2].num), pt2iu(s[3].num),
            pt2iu(s[4].num), pt2iu(s[5].num),
        )
        f.set_extent(area)
